### ISSUE TITLE
Cape vpn

### ISSUE THREAD

**Issue by Solumnant** [kevoreilly/CAPEv2#235]
# Expected Behavior

machine reaches out to the internet when setting VPN 

# Current Behavior

Machine does not have network access when setting VPN

# Failure Information (for bugs)

https://capesandbox.com/analysis/35641/#network

**Comment by doomedraven** [comment]
restarted vpn, maybe was down, i need to improve the checking

**Comment by demo-god** [comment]
https://github.com/kevoreilly/CAPEv2/blob/9b045366ebfcab678c202290dc194ee10bedf673/lib/cuckoo/core/scheduler.py#L555
This line of code disables vpn everytime an analysis is completed, causing the vpn issue. I was able to remove this if loop and vpn network is working fine for me.
Cape only needs to remove routing table entries and iptables entries which happens on line:
https://github.com/kevoreilly/CAPEv2/blob/9b045366ebfcab678c202290dc194ee10bedf673/lib/cuckoo/core/scheduler.py#L551

**Comment by doomedraven** [comment]
well that depends how the vpn configuration is done, i never had issue and i have cape with vpn for idk like 3y

**Comment by demo-god** [comment]
ah! In that case, do you mind sharing your vpn config with me so that my vpn doesn't disable every time an analysis finishes? Thank you in advance :)

**Comment by doomedraven** [comment]
hello you don't need to do nothign with .ovpn files, it just how you start the vpn,check this https://github.com/kevoreilly/CAPEv2/blob/master/utils/route.py  
then you have also this script that prepares all your vpns and configs to write, it just print and you will need to add that content https://github.com/kevoreilly/CAPEv2/blob/master/utils/vpn2cape.py

and then run as it shows `sudo openvpn --config X.ovpn --script-security 2 --route-noexec --route-up utils/route.py &`

**Comment by doomedraven** [comment]
the main problem that we don't have right now any decent way to keep an eye that if vpn is up or not, as some time interface is up but it doesn't work, so we need some kind of check that check the exit ip and if that isn't correct then restart service/process, but i dn't have time to do it

**Comment by demo-god** [comment]
Ah, I see! So is vpn2cape integrated with cape at the moment and where is it being called? or is it something you run manually to setup VMs before running cape?
So I added the following line to my config and it seems to be working for me, atleast it did over the weekend. Every-time my vpn went down it would restart. Need to leave it on for a longer period of time to be sure.
auth-retry interact
My setup is a little different than yours I believe, correct me if I am wrong. I have vpn tunnels running with routes per routing table per vpn. Upon sample submission, cape only adds "ip rule <vm ip> lookup <vpn tunnel>" and of-course iptable rules.

**Comment by doomedraven** [comment]
you run it manually to prepare all the configs as you need to fill another confs that are not related to CAPE, but to linux

well so in your setup the problem is next, what if you want to route 2 VMs over the same vpn? if you have all setup, cape rooter does the rest for vms

**Comment by demo-god** [comment]
Yes, that is true. I am having that issue at the moment. 2 vms cannot route through the same vpn at the same time.
I ll try and replicate your vpn setup, once I have that, I can spend some time working on the script to check if vpn is working.

**Comment by doomedraven** [comment]
just follow the output from that vpn2cape and you will have working vpn

**Comment by demo-god** [comment]
my current configs and a python script to monitor vpn tunnels
https://github.com/demo-god/cape-scripts/tree/master/openvpn

**Comment by doomedraven** [comment]
Thanks i will try to check once i 2ill have some time

**Comment by doomedraven** [comment]
thanks there is similar one already https://github.com/kevoreilly/CAPEv2/blob/master/utils/vpncheck.py

**Comment by koubadomik** [comment]
Sorry, maybe bad question. But it is related to this topis.
I did:
1. create .ovpn file
2. used vpn2cape.sh to generate configuration which I imported to routing.conf and enabled VPN
3. run rooter as root (as referenced in documentation` sudo python3 /opt/CAPEv2/utils/rooter.py --iptables /usr/sbin/iptables --iptables-restore /usr/sbin/iptables-restore --iptables-save /usr/sbin/iptables-save -g cape`
4.  `sudo openvpn --config /etc/openvpn/vpn0.ovpn --script-security 2 --route-noexec --route-up /opt/CAPEv2/utils/route.py` causes: 
```
Options error: --route-up script fails with '/opt/CAPEv2/utils/route.py': Permission denied (errno=13)
Options error: Please correct this error.
Use --help for more information.

```

I definitely missed something, but it is running in sudo, so I do not know what permissions is it missing.


Btw. if I run cuckoo itself with this config:
```
[routing]
route=vpn1
.
.
.
[vpn]
# By default we disable VPN support as it requires running utils/rooter.py as
# root next to cuckoo.py (which should run as regular user).
enabled = yes

# Comma-separated list of the available VPNs.
vpns = vpn1

[vpn1]
name = /etc/openvpn/vpn0.ovpn
description = vpn0
interface = tun1
rt_table = [IP.ADDRESS.OF.SERVER]


```

Cuckoo is giving: `CuckooStartupError: The VPN/Socks5 defined as default routing target has not been configured in routing.conf.`

What I misconfigured/misunderstand?

Thank you in advance for any tip.

**Comment by doomedraven** [comment]
chmod a+x utils/route.py?

**Comment by koubadomik** [comment]
oh, gosh.... You are right, thank you very much. So now I have to solve only the startup fail.

**Comment by doomedraven** [comment]
what is the startup error?

**Comment by koubadomik** [comment]
if I run cuckoo itself with this config:
```
[routing]
route=vpn1
.
.
.
[vpn]
# By default we disable VPN support as it requires running utils/rooter.py as
# root next to cuckoo.py (which should run as regular user).
enabled = yes

# Comma-separated list of the available VPNs.
vpns = vpn1

[vpn1]
name = /etc/openvpn/vpn0.ovpn
description = vpn0
interface = tun1
rt_table = [IP.ADDRESS.OF.SERVER]
`

Cuckoo is giving: `CuckooStartupError: The VPN/Socks5 defined as default routing target has not been configured in routing.conf.
```
What I misconfigured/misunderstand? Even if I try configure it in vpn.conf. Still the same error.

**Comment by doomedraven** [comment]
it should use name field -> `name = /etc/openvpn/vpn0.ovpn` as routing and idk why you set path to ovpn there

**Comment by koubadomik** [comment]
This was produced by `sudo python3 vpn2cape.py /etc/openvpn/`

```
[vpn_1]
name = /etc/openvpn/vpn0.ovpn
description = vpn0
interface = tun1
rt_table = 147.32.82.210




[+] VPNs for CAPE's routing.conf
vpn_1



[+] rt_table for /etc/iproute2/rt_tables
5 147.32.82.210



[+] Paths to execute all in one
sudo openvpn --config vpn0.ovpn --script-security 2 --route-noexec --route-up utils/route.py &
```

After changing it to name string, cuckoo started. Tomorrow I will try network connectivity from the machine during analyses. Thanks again.

**Comment by doomedraven** [comment]
ah damn, thanks i will need to add note about that

**Comment by koubadomik** [comment]
How do you think I should test it?
I thought that I will just run analysis and go to the running VM gui and try to connect to the internet through VPN. It is not working.   Should it work or I have to find another way of testing that VPN is ok and malware is able to access internet?

Beside analysis I am running:
`sudo openvpn --config /etc/openvpn/vpn0.ovpn --script-security 2 --route-noexec --route-up /opt/CAPEv2/utils/route.py`

and 

`sudo python3 /opt/CAPEv2/utils/rooter.py --iptables /usr/sbin/iptables --iptables-restore /usr/sbin/iptables-restore --iptables-save /usr/sbin/iptables-save -g cape`

**Comment by doomedraven** [comment]
why not just to use `curl` using tunX interface?

**Comment by doomedraven** [comment]
now I will wait for some beer as i need to do your job :D

**Comment by koubadomik** [comment]
Indeed, sorry for asking every time, but before I think about a solution every time.

On the other hand I explored that the original cmd running VPN is giving - `Failed running command (--route-up): could not execute external program` which I missed, so I will solve this first and then use `curl` as you kindly provided :D.

**Comment by koubadomik** [comment]
So I know that the vpn is working and I fixed even privileges issues which I have had with route.py 🤦 

Than I got `ERROR: root is wrong table value`. After some investigation I found that env. variable `common_name` is set to root and in route.py it is used as name of the target table to which rules are added. But I am not sure.

I changed value of this variable to 5, as I have in rt_table file this line:
`5       IP_ADDRESS_OF_VPN_SERVER`

After that two rules are added through route.py to the table 5 when running vpn.
```
Fri Nov 20 12:07:59 2020 /sbin1/ip addr add dev tun1 172.16.1.2/24 broadcast 172.16.1.255
ip rule add from 172.16.1.2 table 5
ip route add default via 172.16.1.1 dev tun1 table 5
```

If I try connection from VM during analysis, it is not able to connect anywhere but to gw. 

I know that you are busy and you do not have time to solve my issues, I will go on with solving it. But just in case. Thanks.
